### ISSUE TITLE
Inserting to DateTime64 with best_effort and specific values returns an error (edge case)

### ISSUE THREAD

**Issue by gyfis** [ClickHouse/ClickHouse#28034]
### Describe the bug

Inserting to a DateTime64 column via `INSERT INTO table (...) VALUES (...)` with `date_time_input_format=best_effort` fails when a specific order of the `VALUES` is input. 

The failing order of VALUES: a string with a millisecond unix timestamp followed by a string with a best-effort-style date tim..

### Does it reproduce on recent release?

Yes (tested on 21.8.4).

### How to reproduce

* Which ClickHouse server version to use: 21.8.4
* Which interface to use, if matters: I don't think it does, but I used clickhouse-client
* Non-default settings, if any: `date_time_input_format=best_effort` input into clickhouse client:

```
docker run -it --rm --link some-clickhouse-server:clickhouse-server yandex/clickhouse-client --host clickhouse-server --date_time_input_format=best_effort
```

* `CREATE TABLE` statements for all tables involved:

```sql
CREATE TABLE test_table (datetime DateTime64(3, 'UTC') DEFAULT NOW('UTC')) ENGINE MergeTree ORDER BY datetime
```

### Queries to run that lead to **expected** result

```sql
INSERT INTO test_table (datetime) VALUES (NULL)
INSERT INTO test_table (datetime) VALUES ('2021-08-01T12:23:34.567Z')
INSERT INTO test_table (datetime) VALUES (1629733517)
INSERT INTO test_table (datetime) VALUES ('1629733517')
INSERT INTO test_table (datetime) VALUES (1629733517000)
INSERT INTO test_table (datetime) VALUES ('1629733517000')
INSERT INTO test_table (datetime) VALUES (NULL), (1629733517), ('2021-08-01T12:23:34.567Z'), ('1629733517'), ('1629733517000')
INSERT INTO test_table (datetime) VALUES ('1629733517'), ('2021-08-01T12:23:34.567Z')
INSERT INTO test_table (datetime) VALUES (1629733517000), ('2021-08-01T12:23:34.567Z')
```

### Queries to run that lead to **unexpected** result

When you want to insert a **millisecond unix timestamp as string** before a string timestamp, the query fails

```sql
INSERT INTO test_table (datetime) VALUES ('1629733517000'), ('2021-08-01T12:23:34.567Z')
```

### Expected behavior

I expected the rows to insert in any of the previous cases.

### Error message and/or stacktrace

> Exception on client:
> Code: 6. DB::Exception: Cannot parse string '2021-08-01T12:23:34.567Z' as DateTime64(3, 'UTC'): syntax error at position 23 (parsed just '2021-08-01T12:23:34.567'): while executing 'FUNCTION CAST(assumeNotNull(_dummy_0) :: 2, 'DateTime64(3, \'UTC\')' :: 1) -> CAST(assumeNotNull(_dummy_0), 'DateTime64(3, \'UTC\')') DateTime64(3, 'UTC') : 4': data for INSERT was parsed from query

### Additional info

This could be related to some DateTime64 handling changed e.g. here https://github.com/ClickHouse/ClickHouse/pull/21053 though I didn't dig into it more. 

Let me know if I can provide more info! Thanks

**Comment by den-crane** [comment]
don't reproduce

```sql

CREATE TABLE test_table (datetime DateTime64(3, 'UTC') DEFAULT NOW('UTC')) ENGINE MergeTree ORDER BY datetime;
set date_time_input_format='best_effort';
INSERT INTO test_table (datetime) VALUES ('1629733517000');
INSERT INTO test_table (datetime) VALUES ('2021-08-01T12:23:34.567Z');
INSERT INTO test_table (datetime) VALUES (1629733517.000);

SELECT *
FROM test_table

┌────────────────datetime─┐
│ 2021-08-23 15:45:17.000 │
└─────────────────────────┘
┌────────────────datetime─┐
│ 2021-08-23 15:45:17.000 │
└─────────────────────────┘
┌────────────────datetime─┐
│ 2021-08-01 12:23:34.567 │
└─────────────────────────┘
```

**Comment by gyfis** [comment]
@den-crane Please re-open. You need to insert multiple values, e.g. 

```
INSERT INTO test_table (datetime) VALUES ('1629733517000'), ('2021-08-01T12:23:34.567Z')
```

I listed the _working_ examples on top.

**Comment by den-crane** [comment]
>INSERT INTO test_table (datetime) VALUES ('1629733517000'), ('2021-08-01T12:23:34.567Z')

This should not work. Stream SQL parser deduce format from the first occurrence and uses this format for all following occurrences. It improves performance 1000 times.

**Comment by gyfis** [comment]
@den-crane That doesn't sound 100% correct based on what I see, e.g.

this works
```
INSERT INTO test_table (datetime) VALUES ('2021-08-01T12:23:34.567Z'), ('1629733517000')
```

but this doesn't

```
INSERT INTO test_table (datetime) VALUES ('2021-08-01T12:23:34.567Z'), ('1629733517000'), ('2021-08-01T12:23:34.567Z')
```

If the SQL parser decided based on the initial value, the second example should've worked, too, if I understand you correctly

**Comment by den-crane** [comment]
Datetime & Datetime64 completely different types with own not related behaviors.

Workaround

```
INSERT INTO test_table (datetime) SELECT if(length(t)>13, parseDateTime64BestEffortOrZero(t,3), toDateTime64OrZero(t,3)) FROM input('t String') FORMAT Values ('1629733517000'), ('2021-08-01T12:23:34.567Z');

select * from test_table;
┌────────────────datetime─┐
│ 2021-08-01 12:23:34.567 │
│ 2021-08-23 15:45:17.000 │
└─────────────────────────┘
```

**Comment by den-crane** [comment]
@gyfis I reopened this. But it will be never fixed, trust me. So I suggest to clean your data before loading into CH.

**Comment by gyfis** [comment]
@den-crane Gotcha (even though I still think this behavior is not expected to happen :)). I'm only concerned about DateTime64(3). Thank you for the workaround! Internally we settled on sorting the data before constructing the INSERT query (keeping the long unix timestamps in strings last).

**Comment by den-crane** [comment]
DateTime64 is based on Decimal format. 
CH expects . (period) in  DateTime64 values for milliseconds `INSERT INTO test_table (datetime) VALUES (1629733517.000);` or  text format `'2021-08-01T12:23:34.567`

**Comment by amosbird** [comment]
> @den-crane Gotcha (even though I still think this behavior is not expected to happen :)). I'm only concerned about DateTime64(3). Thank you for the workaround! Internally we settled on sorting the data before constructing the INSERT query (keeping the long unix timestamps in strings last).

It's definitely an unexpected behavior.  `date_time_input_format=best_effort` is not used in the internal CAST function and the best effort parser is not able to parse 13 digits (however, normal parser can)

```
select parseDateTime64BestEffort('1629733517000')  -- error
select toDateTime64('1629733517000', 3) -- ok
```

We either need to improve the best effort parser or make CAST aware of `date_time_input_format`.

**Comment by gyfis** [comment]
I believe this was implemented in https://github.com/ClickHouse/ClickHouse/pull/33276/files - are you okay with closing this one @den-crane ?

**Comment by gyfis** [comment]
Hi @amosbird 👋 I'm confirming that the original issue doesn't happen anymore (tested on v22.5) - most likely fixed in #33276. Closing the issue! Thanks

**Comment by pkit** [comment]
Will need to reopen:
```sql
CREATE TABLE dt64
(
    `timestamp` DateTime64(3)
)
ENGINE = TinyLog

insert into dt64 values (1703363853.035)

INSERT INTO dt64 FORMAT Values

Query id: ce27a0b4-1b3e-42e2-bd78-9b5a208869c6

Ok.

1 row in set. Elapsed: 0.006 sec. 

insert into dt64 format JSONEachRow {"timestamp":1703363853.035} settings date_time_input_format=best_effort

INSERT INTO dt64 FORMAT JSONEachRow

Query id: 6f8ceb7e-3c06-4ce4-97f1-21e22ea44922

Ok.
Exception on client:
Code: 27. DB::ParsingException: Cannot parse input: expected ',' before: '.035} settings date_time_input_format=best_effort': While executing ParallelParsingBlockInputFormat: data for INSERT was parsed from query: (at row 1)
. (CANNOT_PARSE_INPUT_ASSERTION_FAILED)

```
As you can see it "magically" works for `Values` format but fails if it's `JSONEachRow`
